### PR TITLE
coll: nhb_alltoallw

### PR DESCRIPTION
**Pull request by jain-surabhi-23**
Implement neighborhood Alltoallw using transport (TSP) API

This is the 2nd PR in the series of the PRs for neighborhood collectives - #3371 (tests), #3372 (alltoallw), #3375 (allgather), #3376 (allgatherv), #3377 (alltoall), #3378 (alltoallv).

### CODE REVIEW

File: src/mpi/coll/ineighbor_alltoallw/ineighbor_alltoallw.c [hunk 1]
```diff
@@ -19,8 +19,9 @@
       scope       : MPI_T_SCOPE_ALL_EQ
       description : |-
         Variable to select ineighbor_alltoallw algorithm
-        auto   - Internal algorithm selection
-        linear - Force linear algorithm
+        auto            - Internal algorithm selection
+        linear          - Force linear algorithm
+        gentran_linear  - Force generic transport based linear algorithm
```
Comment (hzhou) [inline src/mpi/coll/ineighbor_alltoallw/ineighbor_alltoallw.c]: Also need add case to `MPIR_Ineighbor_alltoallw_impl` to the switch of `MPIR_Ineighbor_alltoallw_intra_algo_choice`.

File: src/mpi/coll/ineighbor_alltoallw/ineighbor_alltoallw_tsp_linear_algos.h [hunk 2]
```diff
@@ -0,0 +1,126 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil ; -*- */
+/*
+ *  (C) 2006 by Argonne National Laboratory.
```
Comment (hzhou) [inline src/mpi/coll/ineighbor_alltoallw/ineighbor_alltoallw_tsp_linear_algos.h]: Please fix update the year

File: src/mpi/coll/ineighbor_alltoallw/ineighbor_alltoallw_tsp_linear_algos.h [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+                                                      MPIR_TSP_sched_t * sched)
+{
+    int mpi_errno = MPI_SUCCESS;
+    int indegree, outdegree, weighted;
+    int k, l;
+    int *srcs, *dsts;
+    int tag;
+    MPIR_CHKLMEM_DECL(2);
+
+    mpi_errno = MPIR_Topo_canon_nhb_count(comm_ptr, &indegree, &outdegree, &weighted);
```
Comment (hzhou) [inline src/mpi/coll/ineighbor_alltoallw/ineighbor_alltoallw_tsp_linear_algos.h]: Missing `MPIR_FUNC_VERBOSE_ENTER` etc.

**Comment by hzhou** [comment]
@jain-surabhi-23 I added a few review notes. These notes applies to the rest of the `Nhb` PRs as well. 

* Make sure the new files have the correct year -- 2018 or 1019, but not 2006 
* Please rebase the PR to current master. In fact, try rebase each PR to current master instead of based on top of each other. PRs may get rejected or merged out-of-order, which won't work here. If you insist all the PRs in the series have to be accepted together, then merge them into a single PR (which I don't think it is necessary here).
* Make sure the switch case in `MPIR_..._impl` is added for the algorithm choice. 
* Please add logging macros, `MPIR_FUNC_VERBOSE_ENTER` etc. to `..._algos.h`

**Comment by hzhou** [comment]
test:jenkins/ch3/most

**Comment by hzhou** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by jain-surabhi-23** [comment]
@hzhou I am working on creating a single PR for all the neighborhood collectives. I will address your review comments there as well. Not having the switch case for these new algos was a good catch. I am fixing it now. You can cancel the tests meanwhile because they wont catch anything without fixing the switch case.

**Comment by hzhou** [comment]
@jain-surabhi-23 why don't you start working the next sets of PRs in this series. I force pushed some changes into your branch, let's see how the tests goes.

**Comment by jain-surabhi-23** [comment]
@hzhou Actually I am consolidating all the PRs in this series, so that you will have to review only 1 PR. I will let you know when it is done and I will close the rest of the PRs. What are the changes you forced pushed? Anything more than you pointed in the review?

**Comment by hzhou** [comment]
Just a rebase to master and adding the selection cases. Seems the tests all failed, so I must missed something anyway. Go ahead with your new consolidating PRs.

**Comment by hzhou** [comment]
Since the linear algorithms works for both intra comm and inter comm, you may also need to add the inter-comm CVARs.

**Comment by jain-surabhi-23** [comment]
I will add those. Thanks for pointing out!

**Comment by jain-surabhi-23** [comment]
This series of PRs have been consolidated into #3593

**Pull request closed without merging**